domains/pallets/messenger/src/migrations.rs

```diff
@@ -1,203 +1,213 @@
-        inbox_responses.into_iter().for_each(|(key, msg)| {
-            // we do one read from the old storage
-            reads += 1;
 
-            // we do one write to old storage and one write to new storage
-            writes += 2;
-
-            InboxResponsesNew::<T>::insert(key, msg);
-        });
+        reads += 1;
+        let Some(mut weight_tags) = MessageWeightTags::<T>::get() else {
+            // nothing to process, exit
+            return (reads, writes);
+        };
 
-        let outbox = Outbox::<T>::drain().collect::<Vec<_>>();
-        let mut outbox_count = BTreeMap::new();
-        outbox.into_iter().for_each(|(key, msg)| {
-            // we do one read from the old storage
-            reads += 1;
+        // migrate outbox weight tags
+        while writes < count {
+            match weight_tags.outbox.pop_first() {
+                // no more weight tags in outbox
+                None => break,
+                Some(((chain_id, message_id), weight_tag)) => {
+                    OutboxMessageWeightTags::<T>::insert((chain_id, message_id), weight_tag);
+                    writes += 1;
+                }
+            }
+        }
+
+        // migrate inbox responses weight tags
+        while writes < count {
+            match weight_tags.inbox_responses.pop_first() {
+                // no more weight tags in outbox
+                None => break,
+                Some(((chain_id, message_id), weight_tag)) => {
+                    InboxResponseMessageWeightTags::<T>::insert((chain_id, message_id), weight_tag);
+                    writes += 1;
+                }
+            }
+        }
+
+        writes += 1;
+        if weight_tags.outbox.is_empty() && weight_tags.inbox_responses.is_empty() {
+            MessageWeightTags::<T>::kill();
+        } else {
+            MessageWeightTags::<T>::set(Some(weight_tags));
+        }
+
+        (reads, writes)
+    }
 
-            // we do one write to old storage and one write to new storage
-            writes += 2;
+    pub(crate) fn remove_inbox_response_weight_tags<T: Config>(tags: Vec<(ChainId, MessageId)>) {
+        let mut maybe_weight_tags = MessageWeightTags::<T>::take();
+        tags.into_iter().for_each(|(chain_id, message_id)| {
+            InboxResponseMessageWeightTags::<T>::remove((chain_id, message_id));
+            if let Some(weight_tags) = maybe_weight_tags.as_mut() {
+                weight_tags.inbox_responses.remove(&(chain_id, message_id));
+            }
+        });
+        MessageWeightTags::<T>::set(maybe_weight_tags);
+    }
 
-            // total outbox count
-            outbox_count
-                .entry((key.0, key.1))
-                .and_modify(|count| *count += 1)
-                .or_insert(1);
+    pub(crate) fn remove_outbox_weight_tag<T: Config>(tag: (ChainId, MessageId)) {
+        if OutboxMessageWeightTags::<T>::contains_key(tag) {
+            OutboxMessageWeightTags::<T>::remove(tag);
```

Comment:
How about `OutboxMessageWeightTags::<T>::take`

Reply:
I dont need the value here. Removing would avoid decoding the value while take would load the value while removing it.

Would prefer remove here